File: 1714-sum-of-special-evenly-spaced-elements-in-array/1714-sum-of-special-evenly-spaced-elements-in-array.py
class Solution:
    def solve(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        
        n = len(nums)
        dividingPoint = int(sqrt(n))
        mod = 10 ** 9 + 7
        
        prefix = {}
        for y in range(1, dividingPoint+1):
            for startingPoint in range(y):
                prefix[(startingPoint, y)] = [0]
                sm = 0
                x = startingPoint
                while x < n:
                    sm += nums[x]
                    sm %= mod
                    prefix[(startingPoint, y)].append(sm)
                    x += y
                    
                    
        res = []
        for x, y in queries:
            if y > dividingPoint:
                sm = 0
                while x < n:
                    sm += nums[x]
                    sm %= mod
                    x += y
                res.append(sm)
            else:
                startingPoint = x % y
                res.append((prefix[(startingPoint, y)][-1] - prefix[(startingPoint, y)][x//y]) % mod)
        return res

File: 1714-sum-of-special-evenly-spaced-elements-in-array/1714-sum-of-special-evenly-spaced-elements-in-array.java
class Solution {
    public int[] solve(int[] nums, int[][] queries) {
        int n = nums.length;
        int margin = (int)Math.sqrt(n);
        int mod = (int)1e9 + 7;
        long[][] sufsum = new long[n][margin + 1];
        for (long[] suf : sufsum)
            Arrays.fill(suf, -1);

        for (int j = margin; j >= 1; j--)   {
            for (int i = n - 1; i >= 0; i--) {
                sufsum[i][j] = i == n - 1 || i + j >= n ? nums[i] : nums[i] + sufsum[i + j][j];
            }
        }
        int m = queries.length;
        int[] ans = new int[m];
        
        for (int i = 0; i < m; i++) {
            int x = queries[i][0];
            int y = queries[i][1];
            
            if (y > margin) {
                long sum = 0;
                for (int j = x; j < n; j += y)
                    sum += nums[j];
                ans[i] = (int)(sum % mod);
            }
            else    {
                ans[i] = (int)(sufsum[x][y] % mod);
            }
        }
        
        return ans;
    }
}